### PR TITLE
Array Docs and Types

### PR DESCRIPTION
**Pull request by wontruefree**
This adds some docs and examples for methods that did not have some and adds some types.  I left some types empty based on my preserved usage.

### CODE REVIEW

File: src/array.cr [hunk 1]
```diff
@@ -2058,7 +2286,14 @@ class Array(T)
     self
   end
 
-  def update(index : Int)
+  # Modifies *self* at *index* to the value *yieled* block and returns the result of the block.
+  # Raises IndexError if the index is out or range.
```
Comment (jhass) [inline src/array.cr:2290]:
```suggestion
  # Sets the value returned from the given block at the given index.
 #
  # Raises `IndexError` if the index is out or range.
```

File: src/array.cr [hunk 2]
```diff
@@ -2104,6 +2339,7 @@ class Array(T)
     super
   end
 
+  # :nodoc:
   private class PermutationIterator(T)
```
Comment (jhass) [inline src/array.cr:2343]: It's private, it already doesn't show up in docs.

File: src/array.cr [hunk 3]
```diff
@@ -2163,6 +2399,7 @@ class Array(T)
     end
   end
 
+  # :nodoc:
   private class CombinationIterator(T)
```
Comment (jhass) [inline src/array.cr:2403]: Ditto

File: src/array.cr [hunk 4]
```diff
@@ -2224,6 +2461,7 @@ class Array(T)
     end
   end
 
+  # :nodoc:
   private class RepeatedCombinationIterator(T)
```
Comment (jhass) [inline src/array.cr:2465]: Ditto.

File: src/array.cr [hunk 5]
```diff
@@ -2282,14 +2520,15 @@ class Array(T)
     end
   end
 
+  # :nodoc:
   private struct FlattenHelper(T)
```
Comment (jhass) [inline src/array.cr:2524]: Ditto.

File: src/array.cr [hunk 6]
```diff
@@ -2282,14 +2520,15 @@ class Array(T)
...
-    def self.flatten(ary)
+    def self.flatten(ary) : Array(T)
```
Comment (jhass) [inline src/array.cr:2525]: Given this is an internal and undocumented method on an internal and undocumented type, maybe the return type restriction is of little value here?

File: src/array.cr [hunk 7]
```diff
@@ -2282,14 +2520,15 @@ class Array(T)
...
       result = [] of T
       flatten ary, result
       result
     end
 
-    def self.flatten(ary : Array, result)
+    def self.flatten(ary : Array, result) : Array(T)?
```
Comment (jhass) [inline src/array.cr:2531]: Ditto.

File: src/array.cr [hunk 8]
```diff
@@ -1393,10 +1393,10 @@ class Array(T)
   # arr = [1, 2, 3]
   # arr.repeated_combinations(0) # => [[]]
   # arr.repeated_combinations(2) # => [[1, 1], [1, 2], [1, 3], [2, 2], [2, 3], [3, 3]]
-  # arr.repeated_combinations    # => [[1, 1, 1], [1, 1, 2], [1, 1, 3], [1, 2, 2], [1, 2, 3],
-  #                              #     [1, 3, 3], [2, 2, 2], [2, 2, 3], [2, 3, 3], [3, 3, 3]
+  # arr.repeated_combinations    # => [[1, 1, 1], [1, 1, 2], [1, 1, 3], [1, 2, 2], [1, 2, 3], [1, 3, 3], [2, 2, 2] ...]
```
[the remark this answers is not in the document]
Comment (wontruefree) [reply src/array.cr:1396]: The full result was very wide but I can add it back.

[the remark this answers is not in the document]
Comment (jhass) [reply src/array.cr:1396]: Maybe we can add some newlines?

[the remark this answers is not in the document]
Comment (wontruefree) [reply src/array.cr:1396]: this new lines looked a little weird after the formatter ran

[the remark this answers is not in the document]
Comment (jhass) [reply src/array.cr:1396]: Might be something to optimize in the formatter then. But I think removing information here is not helpful to understanding this method.

[the remark this answers is not in the document]
Comment (straight-shoota) [reply src/array.cr:1396]: The code should work with the example checker. So abbreviating is not a solution. Can we use a shorter example?

File: src/array.cr [hunk 9]
```diff
@@ -1393,10 +1393,10 @@ class Array(T)
...
   # ```
-  def repeated_combinations(size : Int = self.size) forall Array(T)
+  def repeated_combinations(size : Int = self.size) forall Array
+    (T)
```
[the remark this answers is not in the document]
Comment (wontruefree) [reply src/array.cr:1399]: The formatter is actually doing this???

**Comment by straight-shoota** [comment]
@wontruefree Is there a particular reason to close this PR? Or do you just don't intend to put any more effort in it?

**Comment by wontruefree** [comment]
@straight-shoota I would like to work on these Docs PRs.  There were a lot of comments and some of them I was confused on where they stood.  I figure I would just submit smaller PRs that could be reviewed more easily.

**Pull request closed without merging**